File: java_hummus/src/main/java/com/github/klee0kai/hummus/model/Group.java
package com.github.klee0kai.hummus.model;

import java.util.List;
import java.util.Objects;

public class Group<K, T> {

    public K key = null;

    public List<T> items = null;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Group<?, ?> group = (Group<?, ?>) o;
        return Objects.equals(key, group.key) && Objects.equals(items, group.items);
    }

    @Override
    public int hashCode() {
        return Objects.hash(key, items);
    }
}

File: hummus/src/commonMain/kotlin/com/github/klee0kai/hummus/model/Group.kt
package com.github.klee0kai.hummus.model

/**
 * Data class representing a group of items with a shared key.
 *
 * Result of grouping operations where items are organized by a common key.
 * Provides a simple container for a key and its associated items.
 *
 * **Usage example:**
 * ```kotlin
 * data class User(val id: Long, val department: String, val name: String)
 *
 * val users = listOf(
 *     User(1, "Sales", "Alice"),
 *     User(2, "Sales", "Bob"),
 *     User(3, "IT", "Charlie"),
 * )
 *
 * val grouped = users
 *     .groupBy { it.department }
 *     .map { (dept, userList) -> Group(dept, userList) }
 *
 * // Result:
 * // Group("Sales", [User(1, ...), User(2, ...)])
 * // Group("IT", [User(3, ...)])
 * ```
 *
 * Common grouping patterns:
 * ```kotlin
 * // Using the custom group() extension
 * val groups = users.group(
 *     groupId = { it.department },
 *     grouping = { key, items -> listOf(Group(key, items.toList())) }
 * )
 * ```
 *
 * @param K the type of the grouping key
 * @param T the type of items in the group
 * @property key the grouping key shared by all items
 * @property items the list of items with this key
 */
data class Group<K, T>(
    val key: K,
    val items: List<T>,
)
